power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -271,9 +312,32 @@ inline auto get_edge_weights(TransformerGraph const& graph) -> TrafoGraphEdgePro
         if (graph[e].regulated_idx == unregulated_idx) {
             continue;
         }
-        auto edge_res = std::min(vertex_distances[boost::source(e, graph)], vertex_distances[boost::target(e, graph)]);
+        // control_side, two winding or three winding, is always parallel to the
+        // source side of the transformer edge
+        auto const edge_src_rank = vertex_distances[boost::source(e, graph)];
+        auto const edge_tgt_rank = vertex_distances[boost::target(e, graph)];
+        auto const edge_res = std::min(edge_src_rank, edge_tgt_rank);
+
+        // New edge logic for ranking
+        // |  Tap  | Control |         All edges       |
+        // ---------------------------------------------
+        // |   A   |    A    | [B->A], [C->A], [B<->C] |
+        // |   A   |    B    | [A->B], [A->C], [B<->C] |
+        // |   A   |    C    | [A->B], [A->C], [B<->C] |
+        // |   B   |    A    | [B->A], [C<->A], [B->C] |
+        // |   B   |    B    | [A->B], [C<->A], [C->B] |
+        // |   B   |    C    | [B->A], [C<->A], [B->C] |
+        // |   C   |    A    | [A<->B], [C->A], [C->B] |
+        // |   C   |    B    | [A<->B], [C->A], [C->B] |
+        // |   C   |    C    | [A<->B], [A->C], [A->B] |
+        // In two winding trafo, the edge is always pointing to the control side; in three winding trafo edges, the
+        // edges are always pointing parallel to the control side: meaning that for delta configuration ABC, the above
+        // situations can happen.
+        if (edge_src_rank != edge_tgt_rank - 1) {
+            throw AutomaticTapInputError("Transformer is unrechable: control side rank mismatch");
```

Comment:
Why not use the InvalidRegulatedObject? With a message like this?
(Also the terms "unreachable", "rank mismatch" would not be understood by the user).
Thinking of a better message, suggest if you have.

Reply:
`InvalidRegulatedObject` indicates a different type of error, whereas in this case it's the whole configuration of the connections that lead to this error. I agree that we can document something and make reference to that in the error message. Do you have any suggestions for the error message?

Reply:
"Control side must be the side closest to the source" or something along those lines? Is that how I should interpret this edge case?

Reply:
> "Control side must be the side closest to the source" or something along those lines? Is that how I should interpret this edge case?

I think this is the one!

Reply:
nice 😄 if possible, please explicitly mention that "closest" is weighted distance across regulated transformers or that the user should have a look at the documentation what "closest" means.

Reply:
Not sure if that is necessary; the transformer graph is an intermediate step that only we internally do. I think the mentioning of 'relatively closer a source' is already clear enough and won't drag users into unnecessary confusion.

Reply:
Addressed.